shared/triage_metadata.go

```diff
@@ -0,0 +1,257 @@
+// Copyright 2019 The WPT Dashboard Project. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+//go:generate mockgen -destination sharedtest/triage_metadata_mock.go -package sharedtest github.com/web-platform-tests/wpt.fyi/shared TriageMetadataInterface
+
+package shared
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/google/go-github/v28/github"
+	"gopkg.in/yaml.v2"
+)
+
+// TriageMetadataInterface encapsulates triage() method for testing purpose.
+type TriageMetadataInterface interface {
+	Triage(metadata MetadataResults) error
+}
+
+type triageMetadata struct {
+	ctx context.Context
+	metadataGithub
+	logger     Logger
+	httpClient *http.Client
+}
+
+type metadataGithub struct {
+	githubClient *github.Client
+	authorName   string
+	authorEmail  string
+	wptmetadataGitHubInfo
+}
+
+type wptmetadataGitHubInfo struct {
+	sourceOwner   string
+	sourceRepo    string
+	commitMessage string
+	commitBranch  string
+	baseBranch    string
+	prRepoOwner   string
+	prRepo        string
+	prBranch      string
+	prSubject     string
+	prDescription string
+}
+
+func getNewCommitBranchName(ctx context.Context, client *github.Client, sourceOwner string, sourceRepo string) string {
+	commitBranch := "auto-triage-branch" + generateRandomInt()
+	// If the commitBranch name already exist, genereate a new one. We consider that an error means the branch has not been found and needs to
+	// be created.
+	for ref, _, err := client.Git.GetRef(ctx, sourceOwner, sourceRepo, "refs/heads/"+commitBranch); err == nil && ref != nil; {
+		commitBranch = "auto-triage-branch" + generateRandomInt()
+	}
+
+	return commitBranch
+}
+
+func getWptmetadataGitHubInfo(ctx context.Context, client *github.Client) wptmetadataGitHubInfo {
```

Comment:
I'll leave this up to you, but a bunch of this is duplicated constants (which I don't see changing) for every query that comes in. I would personally have gone for:

```
const (
    sourceOwner = "web-platform-tests"
    sourceRepo = "wpt-metadata"
    baseBranch = "master"
    commitMessage = "Commit New Metadata"
    prSubject = "Automatically Triage New Metadata"
    prDescription = "PR for metadata triaged through /api/metadata/triage endpoint. See <insert a doc> for more information about how to use this service."
)
```

And then `wptmetadataGitHubInfo` wouldn't exist, `metadataGithub` would just have a `commitBranch` member.

Reply:
Right I would have made the same choice, besides there is one annoying hiccup - constant variables are not addressable, [see here](https://play.golang.org/) and [explanation](https://stackoverflow.com/questions/35146286/find-address-of-constant-in-go). I have to create extra variables to indirectly get the address

Reply:
Wow, TIL. And I see that a bunch of the GitHub APIs require pointers (apparently to make distinguishing 'unset' easy: https://godoc.org/github.com/google/go-github/github#hdr-Creating_and_Updating_Resources)

Thanks for the explanation!

Reply:
np. We also use it in the metadata.go to distinguish [unset fields](https://github.com/web-platform-tests/wpt.fyi/blob/master/shared/metadata.go#L47)